python/pyspark/pandas/data_type_ops/binary_ops.py

```diff
@@ -100,7 +99,9 @@ def astype(self, index_ops: IndexOpsLike, dtype: Union[str, type, Dtype]) -> Ind
         if isinstance(dtype, CategoricalDtype):
             return _as_categorical_type(index_ops, dtype, spark_type)
         elif isinstance(spark_type, BooleanType):
-            return _as_bool_type(index_ops, dtype)
+            # Cannot cast binary to boolean in Spark.
+            # We should cast binary to str first, and cast it to boolean
+            return index_ops.astype(str).astype(bool)
```

Comment:
@itholic can you describe the changes that are not test-only in pr description?

Reply:
cc @xinrong-databricks FYI

Reply:
added!

Reply:
can you describe in more details with the input/output before/after this change.

Reply:
btw, why not fixing it in `_as_bool_type`?

Reply:
moved it into `_as_bool_type` and updated the PR description!

Reply:
`_as_bool_type` should not include `index_ops`'s type dependent code.
We should revert the change c429108d380d5dfaf4402d8382bb26d6a0ca32fc.